spring-cloud-starter-stream-source-sftp-dataflow/src/main/java/org/springframework/cloud/stream/app/sftp/dataflow/source/tasklauncher/SftpTaskLaunchRequestArgumentsMapper.java

```diff
@@ -0,0 +1,156 @@
+	}
+
+	private void addLocalFileCommandLineArgs(Collection<String> commandLineArgs , Message<?>
+		message) {
+
+		String localFilePath;
+
+		if (sourceProperties.isListOnly()) {
+			String filename = (String) message.getPayload();
+			localFilePath =
+				FilePathUtils.getLocalFilePath(sourceProperties.getLocalDir().getPath(),
+					filename);
+		}
+		else {
+			localFilePath = ((File) message.getPayload()).getAbsolutePath();
+		}
+
+		commandLineArgs.add(LOCAL_FILE_PATH_PARAM_NAME + "=" + localFilePath);
+	}
+
+	private void addRemoteFileCommandLineArgs(Collection<String> commandLineArgs, Message<?>
+		message) {
+
+		String remoteFilePath = FilePathUtils.getRemoteFilePath(message);
+
+		if (StringUtils.hasText(remoteFilePath)) {
+			commandLineArgs.add(REMOTE_FILE_PATH_PARAM_NAME + "=" + remoteFilePath);
+		}
+		if (this.sourceProperties.isMultiSource()) {
+			Map<String, Object> headers = convertMultiSourceHeaders(listFilesRotator.headers());
+			commandLineArgs.add(
+					String.format("%s=%s", SftpHeaders.SFTP_SELECTED_SERVER_PROPERTY_KEY,
+							headers.get(SftpHeaders.SFTP_SELECTED_SERVER_PROPERTY_KEY)));
+		}
+	}
+
+	private void addSftpConnectionInfoToTaskCommandLineArgs(Collection<String> commandLineArgs) {
+		if (!this.sourceProperties.isMultiSource()) {
+			commandLineArgs.add(
+				String.format("%s=%s", SftpHeaders.SFTP_HOST_PROPERTY_KEY, sourceProperties.getFactory().getHost()));
+			commandLineArgs.add(String.format("%s=%s", SftpHeaders.SFTP_USERNAME_PROPERTY_KEY,
+				sourceProperties.getFactory().getUsername()));
+			commandLineArgs.add(String.format("%s=%s", SftpHeaders.SFTP_PASSWORD_PROPERTY_KEY,
+				sourceProperties.getFactory().getPassword()));
+			commandLineArgs.add(String.format("%s=%s", SftpHeaders.SFTP_PORT_PROPERTY_KEY,
+				String.valueOf(sourceProperties.getFactory().getPort())));
+		}
+		else {
+			Map<String, Object> headers = convertMultiSourceHeaders(listFilesRotator.headers());
+			commandLineArgs.add(
+					String.format("%s=%s", SftpHeaders.SFTP_SELECTED_SERVER_PROPERTY_KEY,
+							headers.get(SftpHeaders.SFTP_SELECTED_SERVER_PROPERTY_KEY)));
+
+			commandLineArgs.add(
+					String.format("%s=%s", SftpHeaders.SFTP_HOST_PROPERTY_KEY,
+							String.valueOf(headers.get(SftpHeaders.SFTP_HOST_PROPERTY_KEY))));
+
+			commandLineArgs.add(
+				String.format("%s=%s", SftpHeaders.SFTP_PORT_PROPERTY_KEY,
+					String.valueOf(headers.get(SftpHeaders.SFTP_PORT_PROPERTY_KEY))));
+
+			commandLineArgs.add(String.format("%s=%s", SftpHeaders.SFTP_USERNAME_PROPERTY_KEY,
+				headers.get(SftpHeaders.SFTP_USERNAME_PROPERTY_KEY)));
+
+			commandLineArgs.add(String.format("%s=%s", SftpHeaders.SFTP_PASSWORD_PROPERTY_KEY,
+				headers.get(SftpHeaders.SFTP_PASSWORD_PROPERTY_KEY)));
+		}
+	}
+
+	private Map<String, Object> convertMultiSourceHeaders(Map<String, Object> headers) {
+		Map<String, Object> result = new HashMap<>();
+		headers.forEach((k, v) -> result.put(k, ((FunctionExpression) v).getValue()));
```

Comment:
Is there really guarantee that header values are always `FunctionExpression` instances?

Reply:
This was just refactoring existing code.  The `SftpSourceRotator.headers()` provides `FunctionExpression` for these values. I don't know the original reason for using FunctionExpression.

Reply:
See my comment in the `SftpSourceRotator` about having that `headers()` method to return a map with `Expression` specialization.